### ISSUE TITLE
Overhaul Power Boost / Overclock ITab 

### ISSUE THREAD

**Issue by Sn1p3rr3c0n** [zymex22/Project-RimFactory-Revived#279]
as Suggested by `Dr_Zhivago` and discussed on the Discord

_ASCII art 1 ~Dr_Zhivago_
```
Power Usage: [700]
Increase base power usage for Work Speed:
0                                           2000
------------------------------------------------
                                       [       ]
=================================================
Increase base power usage for Range:
0                                           2000
------------------------------------------------
                                       [       ]
```
_ASCII art 2 ~Dr_Zhivago_
```
Total Power Usage:
Base [700] + Work Speed [---] + Range [---] = [700]

Increase base power usage for Work Speed:
0                                           2000
x------------------------------------------------

Increase base power usage for Range:
0                                           2000
x------------------------------------------------
```
---

Summary:
- Remove Input Box as Steps are used
- Add divider line as shown in Art 1
- make it so that the boost power Display starts at 0 and ends with its maximum boost
- Add a  Breakdown of the power usage as shown in Art 2

**to be considered**
- Support for other additional Power Drains such as #196 
- Unify ITab for all Buildings #236

**Comment by zymex22** [comment]
maybe also look at #176 if thats related to slider also in some way =)

**Comment by Sn1p3rr3c0n** [comment]
![powerTab_pverhaulv1](https://user-images.githubusercontent.com/68663281/102378013-9d97a380-3fc5-11eb-9108-e45d09a45095.PNG)
Added initial UI Changes with a5e3406

**Comment by Sn1p3rr3c0n** [comment]
Testing shows that all Buildings use the same Power Slider ITab. 
Seems like #236 is either for the Unification of XML Settings or the removal of dead code

**Comment by Sn1p3rr3c0n** [comment]
Current todo list
- [ ] Fully port pullers over
- [ ] Remove old ModExtension 
- [ ] fix bugs
- [ ] Fix XML Configuration
- [ ] Cleanup lots of C#
- [ ] Get feedback & Improve UI
- [ ] fix more bugs?
- [ ] maybe check #176

**Comment by Sn1p3rr3c0n** [comment]
Preview of the Updated look ca8e64d
![powerTab_pverhaulv2](https://user-images.githubusercontent.com/68663281/102711044-14c78300-42b7-11eb-99f6-73a30fa6b1a5.PNG)

**Comment by Sn1p3rr3c0n** [comment]
**Known issues:**
- the functionality described in #310 has been remove in this version. _(we can still add it back for all with #310 but we should not increase the size of this overhaul even more)_

**Comment by zymex22** [comment]
After some initial testing it looking pretty promising! :D

- those buildings that used AutoMachineTool speed settings before eq. belt WILL need a update :P speed system are entirely different from what it was before...
(this can be tweaked by me later)

- See my suggestion on Discord for Speed / Range info
![image](https://user-images.githubusercontent.com/41029070/102716975-9b905600-42df-11eb-9697-82fc67071ab6.png)


- Dunno if this is Related here? for Settings Itab.
but maybe we should update that Output part of it:
![image](https://user-images.githubusercontent.com/41029070/102716953-6dab1180-42df-11eb-9404-c0daa7ccbef2.png)
so remove line between those 2 settings as the control same thing.. 
and a Header that say whatever those options actually do control

- Can T3 miner and other ones that didnt have settings Itab before have that now as well?
(it dont now) but can it? i seet speed works on it =)

- is this related or resolve or should resolve #154 also ? if so we should ask @lilwhitemouse whats the plan so it can be done from start right?

- found this, one case where sizeing for some reason was a bit funny? :P
![image](https://user-images.githubusercontent.com/41029070/102717163-d646be00-42e0-11eb-80e2-4f6bb43b5916.png)



-- i couldnt find any machines that didnt work or was making funny things, i tried most of the AutomachineTool ones as i suspected those to be worst, as they used old slider system.
though they will need a speed update... but thats easy for my part =)

**Comment by Sn1p3rr3c0n** [comment]
There where no changes to the Settings ITab.
Therefor all change Requests Relating to the Settings ITab shall be documented & fixed in a different issue

**Comment by Sn1p3rr3c0n** [comment]
@zymex22 i would have a proposal to change how the configuration of `ProjectRimFactory.Common.CompProperties_PowerWorkSetting` works.
But as this will only impact the configuration side for XML i would like your opinion on it.

---
###  Current Implementation
**Has the following XML Settings**
`minRange`
`maxRange`
`maxPowerForRange`

`minSpeedFactor`
`maxSpeedFactor`
`maxPowerForSpeed`

**logic**
Power usage per Range step is calculated
`maxPowerForRange / (maxRange - minRange)`
Step Count (Range)
`(maxRange - minRange)`

Power usage per Speed step is
`50`
Step Count (Speed)
`maxPowerForSpeed / 50`

---
### Proposal
**Will have the following XML Settings**
`minRange`
`maxRange`
`powerPerStepRange`

`minSpeedFactor`
`maxSpeedFactor`
`powerPerStepSpeed`
(optional) `powerStepFactor`


**logic**
Power usage per Range step is calculated
`powerPerStepRange`
Step Count (Range)
`(maxRange - minRange)`

Power usage per Speed step is
`powerPerStepSpeed`
Step Count (Speed)
`(maxSpeedFactor - minSpeedFactor)`

(optional) Power usage per Speed step is
`powerPerStepSpeed / powerStepFactor`
(optional) Step Count (Speed)
`(maxSpeedFactor - minSpeedFactor) * powerStepFactor`


---

**Comment by Sn1p3rr3c0n** [comment]
@zymex22 You can try this Version & Provide some feedback.

**Notes / Known issues:**
- if `floatrange_SpeedFactor` or `floatrange_Range` holds a decimal value the Slider doesn't function as expected
- The Range & Speed Display likely needs a User Translation. Thoughts / Ideas @zymex22 ?

**Comment by zymex22** [comment]
- hmmm decimal like 1.5?
it would never really be needed to do that in Range...
but might be in Speed..
i have used 1.5 before for (150% speed)
if any easier we can just use 100=100% instead of 1.0=100%

- I'll do translations / setting up text when i get my hands on it you can just put some basic in it, and i change to something better if needed.


-- I'll test this after work and see how it works =)

**Comment by Sn1p3rr3c0n** [comment]
@zymex22 The decimal issue has been resolved with d7d4609

the issue is that for sliders to work correctly we need a [Natural number](https://en.wikipedia.org/wiki/Natural_number) of steps.

**Example with the issue**
```XML
<li Class="ProjectRimFactory.Common.CompProperties_PowerWorkSetting">
    <!-- Speed Settings -->
    <floatrange_SpeedFactor>1~3.5</floatrange_SpeedFactor>
    <powerPerStepSpeed>4000</powerPerStepSpeed>
    <powerStepFactor>1</powerStepFactor>
    <!-- Range Settings -->
    <floatrange_Range>1~5</floatrange_Range>
    <powerPerStepRange>125</powerPerStepRange>
    <allowManualRangeTypeChange>true</allowManualRangeTypeChange>
    <rangeType>ProjectRimFactory.Common.RectRange</rangeType>
</li>
```

**Explanation & Fix**
As you can see from the example we have `2.5` Steps. This is not a  [Natural number](https://en.wikipedia.org/wiki/Natural_number).
To resolves the issue we can use `powerStepFactor`.
`powerStepFactor` is `1` by default _(must never be 0)_ 
We now need to choose a number for `powerStepFactor` where:
`powerStepFactor * Step-Count = Natural Number`
There are multiple options for this, in d7d4609 i chose `10` but `2` would also work just fine.
```
2.5 * 10 = 25 (is a Natural Number)
2.5 * 2 = 5 (is a Natural Number)
```


At this stage it is also important to keep the effect of `powerStepFactor` in mind.
besides resolving our issue it also changes our actual In-Game `Power per step`
the Formula is as follows:
`powerPerStepSpeed / powerStepFactor = In-Game Power per step` 
So the two value from above will have the following effects:
```
4000 / 10 = 400   (In-Game Power per step)
4000 / 2   = 2000 (In-Game Power per step)
```

**Example of Fix**
```XML
<li Class="ProjectRimFactory.Common.CompProperties_PowerWorkSetting">
    <!-- Speed Settings -->
    <floatrange_SpeedFactor>1~3.5</floatrange_SpeedFactor>
    <powerPerStepSpeed>4000</powerPerStepSpeed>
    <powerStepFactor>2</powerStepFactor>
    <!-- Range Settings -->
    <floatrange_Range>1~5</floatrange_Range>
    <powerPerStepRange>125</powerPerStepRange>
    <allowManualRangeTypeChange>true</allowManualRangeTypeChange>
    <rangeType>ProjectRimFactory.Common.RectRange</rangeType>
</li>
```

---
For Range this will never be relevant as there is no point in having a decimal range.
in Rimworld you either include a cell or not. there is no middle ground.

**Comment by zymex22** [comment]
some initial thoughts:
- Speed is in % right? so 1= 100%, 5=500% of GlobalWorkSpeed right?

- if so maybe we should list it as 100% and 500% instead or write x100 % after. like on cars rpm meter :P

- speed middle one should probably say
Current Speed (XX) % Current Radius (XX) Cells

- seems overclock tab have been removed on pullers, need to readd and test there also to see it still works before we release it.

- what happens if using range on a building that didn't use range slider before eq. Puller... :stuck_out_tongue: will it just throw a error and need to fix in XML or something else?

- and using speed on something that didnt support speed before.. i see Spacer Animal Harvester

Actually that one might be broke or i forgot how it works? did it only harvest one wool per turn? it atlease does now.
some initial thoughts:
- Speed is in % right? so 1= 100%, 5=500% of GlobalWorkSpeed right?

- if so maybe we should list it as 100% and 500% instead or write x100 % after. like on cars rpm meter :P

- speed middle one should probably say
Current Speed (XX) % Current Radius (XX) Cells

- seems overclock tab have been removed on pullers, need to readd and test there also to see it still works before we release it.

- what happens if using range on a building that didn't use range slider before eq. Puller... :stuck_out_tongue: will it just throw a error and need to fix in XML or something else?

- and using speed on something that didnt support speed before.. i see Spacer Animal Harvester.
Actually that one might be broke or i forgot how it works? did it only harvest one wool per turn? it atlease does now.
![image](https://user-images.githubusercontent.com/41029070/102780125-b45b4300-4395-11eb-9895-eb9232bb1eb8.png)

- oh another small thing? 
dunno if possible somehow?
but would we be able to list what "range actually does" like if i can set in xml use language line. 1, 2, 3. or something for
`Adjust import range`
`Adjust workzone range`
dunno if theres more cases than that...
but for drones its more like a workzone.
where on assemblers its import/pulling range instead :slight_smile: 
you see what i mean? XD¨'



other than that it looks really good ill do some more testing after i get a shower :slight_smile:
